src/server/world.zig

```diff
@@ -928,6 +935,38 @@ pub const ServerWorld = struct { // MARK: ServerWorld
 		self.dropWithCooldown(stack, pos, dir, velocity, 0);
 	}
 
+	fn tickBlocksInChunk(self: *ServerWorld, _chunk: *chunk.ServerChunk) void {
+		for(0..self.tickSpeed) |_| {
+			const blockIndex: i32 = main.random.nextInt(i32, &main.seed);
+
+			const x: i32 = blockIndex >> chunk.chunkShift2 & chunk.chunkMask;
+			const y: i32 = blockIndex >> chunk.chunkShift & chunk.chunkMask;
+			const z: i32 = blockIndex & chunk.chunkMask;
+
+			_chunk.mutex.lock();
+			const block = _chunk.getBlock(x, y, z);
+			_chunk.mutex.unlock();
+			if(block.tickEvent()) |event| {
+				event.tryRandomTick(block, _chunk, x, y, z);
+			}
+		}
+	}
+
+	fn tick(self: *ServerWorld) void {
+		// tick blocks
+		ChunkManager.mutex.lock();
+		var iter = ChunkManager.entityChunkHashMap.keyIterator();
+		while(iter.next()) |pos| {
+			if(ChunkManager.entityChunkHashMap.get(pos.*)) |entityChunk| {
+				const ch = entityChunk.getChunk() orelse continue;
+				entityChunk.increaseRefCount();
```

Comment:
Alright, so some lessons on multi-threading:
- Doing refcounting while already holding the lock of the parent data structure is pointless. We do refcounting to allow us to keep using it on changes to the parent data structure.
- For refcounting you must increase the refcount while holding the lock to the containing data structure, afterwards you are free to decrease it outside the lock.

And here we don't want to hold the lock while executing the tick function, since as you discovered it causes a deadlock.

I think the correct solution would be to copy all chunks into a list first and increase all of their refcounts while holding the lock, and then afterwards you can actually go through them and execute the tick function without holding any lock.